types/index.d.ts

```diff
@@ -1813,7 +1813,8 @@ declare module 'mongoose' {
 
   type QueryWithHelpers<ResultType, DocType, THelpers = {}, RawDocType = DocType> = Query<ResultType, DocType, THelpers, RawDocType> & THelpers;
 
-  type UnpackedIntersection<T, U> = T extends (infer V)[] ? (V & U)[] : T & U;
+  type UnpackedIntersection<T, U> = T extends (infer A)[] ? (A & U)[] : keyof U extends never ? T & U: { [P in keyof U]: U[P] extends (infer B)[] ? (B & T)[] : U[P] & T };
```

Comment:
Shouldn't the last bit be something like `T[P] & U[P]`? I think `T` should come first, and then we should be merging T's keys as well as opposed to merging T into U's keys.

Reply:
Check the new commit